### DISCUSSION TITLE
nwg-shell -> Arch Extra

### DISCUSSION THREAD

**Discussion by nwg-piotr** [nwg-piotr/nwg-shell#374] [General]
@Antiz96, that's the list you required.

| Package| Action |
| -- | -- |
| azote |  necessary |
| ~~lightdm-nwg-greeter~~ | use nwg-hello instead |
| gopsuinfo | necessary |
| gtklock | foreign, necessary |
| gtklock-* stuff | foreign, to be discussed, possibly not all of them |
| nwg-bar | could be of use, not really necessary |
| nwg-clipman | necessary | 
| nwg-displays | necessary |
| nwg-dock | necessary |
| nwg-dock-hyprland | necessary |
| nwg-drawer | necessary |
| *nwg-icon-picker* | *packaged* |
| *nwg-look* | *packaged* |
| nwg-menu | necessary |
| *nwg-panel* | *packaged* |
| nwg-readme-browser | necessary |
| nwg-shell | necessary |
| nwg-shell-config | necessary |
| nwg-shell-wallpapers | necessary |

**Comment by Antiz96** [comment]
Thanks for the list!
I'm currently working on it and I should be ready to package them soon*ish*.

However, I'm facing an issue: It seems like `nwg-shell` & `nwg-shell-config` are expecting some files to be shipped under `/usr/local/...` (see [this](https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=nwg-shell#n37) & [this](https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=nwg-shell-config#n36)). As per the [Arch packaging guidelines](https://wiki.archlinux.org/title/Arch_package_guidelines#Package_etiquette) (and overall usual packaging standards), a package should never install anything under `/usr/local`. So those files will need to be packaged somewhere else (e.g. `/usr/bin/` or `/usr/share` or something). I assume this implies some upstream code changes?

Once the above issue is solved and I'm ready to build every packages, I'll upload them all to [extra-testing] so we can run some tests. For instance by creating an Arch test environment using the testing repo, install the `nwg-shell` *meta*package from there and check that every nwg-* packages/components are present and working as intended. I'll do it myself to check the packaging side of things but it would be great to have a check from you as well to verify that everything works as expected :)

Once done, I'll move everything to [extra] :rocket: :tada:

**Comment by Antiz96** [comment]
Additionally, once it has entered [extra], I think that an Arch Wiki page dedicated to the `nwg-shell` would be a great plus!
But that can be done in a later time, and we can hopefully count on the community to help on that front :)

**Comment by nwg-piotr** [comment]
>  I assume this implies some upstream code changes?

Yes, and it requires me to be sober and at my own desk, that's gonna be possible next week. I'd also like to make some changes to the `nwg-shell-installer` script.

> I think that an Arch Wiki page dedicated to the nwg-shell would be a great plus!

It's something you never have time for, but I'll do my best.

**Reply by Antiz96** [reply]
> > I assume this implies some upstream code changes?
> 
> Yes, and it requires me to be sober and at my own desk, that's gonna be possible next week. I'd also like to make some changes to the `nwg-shell-installer` script.

Alright, no problem. Thanks :)

> > I think that an Arch Wiki page dedicated to the nwg-shell would be a great plus!
> 
> It's something you never have time for, but I'll do my best.

Haha, fair enough. Again, this is not mandatory, just a nice plus.

**Comment by Antiz96** [comment]
I'm ready to package the following list of software, which should cover the list you gave me and the various dependencies for `nwg-shell` (+ one or two extras):

- autotiling 
- azote 
- gopsuinfo 
- gtk-session-lock 
- gtklock 
- gtklock-playerctl-module 
- gtklock-powerbar-module 
- gtklock-userinfo-module 
- nwg-bar 
- nwg-clipman 
- nwg-displays 
- nwg-dock 
- nwg-dock-hyprland 
- nwg-drawer 
- nwg-hello 
- nwg-menu 
- nwg-readme-browser 
- nwg-shell 
- nwg-shell-config 
- nwg-shell-wallpapers

As shown in your list, the following software are already packaged in [extra]:

- nwg-icon-picker
- nwg-look
- nwg-panel

Tell me if I missed anything :)

Once the changes to avoid using `/usr/local` have been made in `nwg-shell`/`nwg-shell-config`, I'll build and upload everything in [extra-testing] :wink:

**Reply by nwg-piotr** [reply]
Looks good to me, but I'm in the wild, with a 14 y.o. spare lappy, and nothing of my development suite in range. You could just package what's possible at the moment, except for what needs changes. I'll make them asap, when I'm back home.

**Reply by Antiz96** [reply]
Sure, no problem. It was just to confirm the list and give a quick update. 
Actually, I may be mostly AFK until next week as well :)

**Reply by nwg-piotr** [reply]
Very well. We're still still in touch one way or another.

**Comment by nwg-piotr** [comment]
I'm home. Will fix [#48](https://github.com/nwg-piotr/nwg-dock-hyprland/issues/48) quickly, and then make changes you required to nwg-shell and nwg-shell-config. Some other minor issues I noticed on Hyprland v0.42.0 will have to wait.

**Comment by nwg-piotr** [comment]
> However, I'm facing an issue: It seems like nwg-shell & nwg-shell-config are expecting some files to be shipped under /usr/local/... (see [this](https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=nwg-shell#n37) & [this](https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=nwg-shell-config#n36)).

These scripts need to be located anywhere on `#PATH` in both packages. No changes to the code were needed. In just released nwg-shell 0.5.38 and nwg-shell-config 0.5.43,  I only changed a line in install.sh scripts (which I use for installing bypassing the package), to copy them to `/usr/bin` instead of `/usr/local/bin`.

**Comment by nwg-piotr** [comment]
One more thing: I've just realized, that the installer script uses the [baph](https://aur.archlinux.org/packages/baph) AUR helper, as we used to need AUR packages. As for now we need either to add baph to our packages (it would be not bad, as it works very well with my updater script), or to rework the installation script.

**Reply by Antiz96** [reply]
> [...] we need either to add baph to our packages [...]

This is not an option. AUR helpers and/or pacman wrappers are not supported and thus are not allowed to enter [extra] (see  [this](https://wiki.archlinux.org/title/Package_Maintainer_guidelines#Rules_for_packages_entering_the_extra_repository))*; and, of course, a package in [extra] cannot depend on an AUR package*.

However, unless I'm mistaken, the installer script is not used by the `nwg-shell` Arch package so it's not really an issue for the Arch package itself.

**Reply by nwg-piotr** [reply]
The installer needs an update one way or another. I'll work on it this afternoon/evening.

**Reply by Antiz96** [reply]
I have released `nwg-shell` and `nwg-shell-config` in [extra-testing] regardless so we can start some testing.

Correct me if I'm wrong, but if you are referring to [this script](https://github.com/nwg-piotr/nwg-shell/blob/main/install/arch.sh), from my point of view it can now be completely dropped (as `nwg-shell` and all its components are now in [extra]). Installing the `nwg-shell` metapkg should now cover most (if not all) of the steps of this script I think, right?

**Reply by nwg-piotr** [reply]
I mean this script, in connection with the `nwg-shell-installer` python mount point (installer.py).

1. For nwg-shell to work, we need some stuff preinstalled: a file manager, a text editor and a web browser. [**edit:** we also need to have installed sway (mandatory) and hyprland (optional)]. That's what we expect  arch.sh to do. I think we should rename it by the way. Then it runs `nwg-shell-installer`, that installs all default configs and does some pre-configuration, among other things, on the basis of installed file manager, text editor, and web browser.

So, we're going to need another release of the nwg-shell package.

2. If it comes to nwg-shell-config, it contains the update indicator applet, that needs to be able to work w/o AUR helpers from now on, so we need to update it as well.

Sorry I didn't think of this last night, but I guess my brain was still fishing.

3. BTW: have you heard of any fresh bluetooth issues on Arch Linux lately? It stopped working for me suddenly, and I can't get it back to work. My mouse doesn't like it. :/

**Reply by nwg-piotr** [reply]
I copied `install/arch-sh` to `install-nwg-shell.sh` and removed no longer necessary stuff. I've also added some comments. If you please could take a look at it, tell me what you think.

https://github.com/nwg-piotr/nwg-shell/blob/extra/install-nwg-shell.sh

**Reply by Antiz96** [reply]
> I mean this script, in connection with the `nwg-shell-installer` python mount point (installer.py).
> 
>     1. For nwg-shell to work, we need some stuff preinstalled: a file manager, a text editor and a web browser. [**edit:** we also need to have installed sway (mandatory) and hyprland (optional)]. That's what we expect  arch.sh to do. I think we should rename it by the way. 

Sorry if I'm missing something, but from my point of view all necessary packages should/will be installed by the `nwg-shell` metapkg. The same package also "opt-depends" on a browser, a file manager and a text editor; so users are expected to know they need one (if anything, this can also be echoed to users during the installation of the package and documented in a wiki page). Unless I'm missing something, I don't think an additional `arch.sh` script is necessary anymore. If anything is missing from the `nwg-shell` metapkg itself; it's either a miss (package wise) or a documentation matter in my opinion.

> Then it runs `nwg-shell-installer`, that installs all default configs and does some pre-configuration, among other things, on the basis of installed file manager, text editor, and web browser.
> 

We can force the execution of the script after the installation of the package via pacman/makepkg if necessary, but can't we just let users run the `nwg-shell-installer` script after installation themselves if they want/need to? Here again, it feels like a documentation matter rather than anything else (unless I'm missing something?).

> 
> So, we're going to need another release of the nwg-shell package.
> 

From my point of view, we need a wiki page to document the above rather than another release. I could be wrong, but from my understanding, I feel like all that should be necessary is to install the `nwg-shell` metapkg, install a web browser/file manager/text editor (if not done already), and then optionally run the `nwg-shell-installer` to generate the default config *if required* (one could have it's own config already). I personally wouldn't expect the necessity to run an additional/"third-party" (package wise) `arch.sh` script to do so. This simply belongs in a wiki page in my opinion. At least, this is usually the way Arch does things generally speaking. Again, sorry if I'm missing something.

**EDIT:** See my comments on the `install-nwg-shell.sh` script below for more details.

>     2. If it comes to nwg-shell-config, it contains the update indicator applet, that needs to be able to work w/o AUR helpers from now on, so we need to update it as well.
> 

Indeed, *hardly* depending on baph (or any AUR helper) is not possible now that packages are moved to [extra].
The script can still use baph if it is installed (even if the package itself cannot depend on it). Again, this can be documented in a wiki page.

> 
> Sorry I didn't think of this last night, but I guess my brain was still fishing.
> 

Haha, no problem ;)

>     3. BTW: have you heard of any fresh bluetooth issues on Arch Linux lately? It stopped working for me suddenly, and I can't get it back to work. My mouse doesn't like it. :/

Nope, everything works fine on my end :/

**Reply by Antiz96** [reply]
Here are my comments about the `install-nwg-shell.sh` scripts (line 24, 35, 42, 78, 84 & 91): https://gist.github.com/Antiz96/3872f929c998517130bb9ce853e3b076

To sum up, I think the steps of this script are either not necessary anymore, should be handled by the `nwg-shell` metapkg itself or should be part of a documentation (message echoed during installation and/or a wiki page). At least, this is how things would be intended to be regarding "the Arch way".
Additionally, in it's current state, the script is subject to breakages as it offers to install outdated/deleted packages (e.g. atom editor) or AUR packages (e.g. brave-bin) via pacman, which will end up in error. The list of proposed packages would need to be *constantly* updated.

Now, if you still want to keep this script as an alternative installation method, it's obviously up to you. But, from my point of view, this script's purpose should be covered directly by the `nwg-shell`'s PKGBUILD and the documentation. Integrating it to the package somehow is unnecessary, subject to breakages and doesn't follow the Arch "principles"/packaging standard. My personal opinion is to drop that script completely. I remain available if you want to discuss that :)

**Reply by nwg-piotr** [reply]
Let's do it the Arch way, if so. If I didn't forget about something, what we need to add to PKGBUILD is:

- install xdg-user-dirs, sway
- echo about the need to have a file manager, text editor & web browser installed
- echo about optional Hyprland support
- echo about the need to run `nwg-shell-installer -w [-hypr]`

> but can't we just let users run the nwg-shell-installer script after installation themselves if they want/need to?

Nwg-shell is a set of programs and predefined desktop styles, installed by this script. The nwg-shell-config utility won't work w/o them. And hell knows what else.

In this point we could do first tests. I'll modify the updates notification applet in the meantime.

**[edit]** One more thing: it seems we don't have `checkupdates` anywhere, and the applet needs it. Could you add it to nwg-shell-config dependencies?

**Reply by Antiz96** [reply]
> Let's do it the Arch way, if so. If I didn't forget about something, what we need to add to PKGBUILD is:
> 
>     * install xdg-user-dirs, sway
> 
>     * echo about the need to have a file manager, text editor & web browser installed
> 
>     * echo about optional Hyprland support
> 
>     * echo about the need to run `nwg-shell-installer -w [-hypr]`
> 

All done! [Here](https://gitlab.archlinux.org/archlinux/packaging/packages/nwg-shell/-/commit/946a4f41f05b159cf522effcaf57ba77d827388d) & [here](https://gitlab.archlinux.org/archlinux/packaging/packages/nwg-shell/-/commit/5c42bc4f477646a18d1ee5fbd3dc28da28b1e62a).

I also added `nwg-hello` [as a dependency](https://gitlab.archlinux.org/archlinux/packaging/packages/nwg-shell/-/commit/f2aec701e80ac071f21fcb97851e9ec78533d562), I guess it makes sense to have it installed with the rest of nwg components when installing the metapkg.

> 
> > but can't we just let users run the nwg-shell-installer script after installation themselves if they want/need to?
> 
> Nwg-shell is a set of programs and predefined desktop styles, installed by this script. The nwg-shell-config utility won't work w/o them. And hell knows what else.
> 
> In this point we could do first tests. I'll modify the updates notification applet in the meantime.
> 

Alright, thanks for the info!

I also created an Arch wiki page for nwg-shell at: https://wiki.archlinux.org/title/Nwg-shell
It's probably missing some relevant information but it should be a good base to start things off. I added it to the [post_install message](https://gitlab.archlinux.org/archlinux/packaging/packages/nwg-shell/-/commit/12667cc0e757f8f5d651bbeff5fdfe1d88884765) for users.
Feel free to add/modify/correct/remove anything from there (or give me the necessary changes if you want me to do it) :smile: 

**[edit]** I also added a note about enabling `greetd.service` to the [post_install message](https://gitlab.archlinux.org/archlinux/packaging/packages/nwg-shell/-/commit/2983eba5fe98a0f62d0048b0e919e5f68a9292b3) for users (this is also covered in the wiki page).

> **[edit]** One more thing: it seems we don't have `checkupdates` anywhere, and the applet needs it. Could you add it to nwg-shell-config dependencies?

`pacman-contrib` (which provides `checkupdates`) was noted as a dependency for [nwg-shell in the AUR](https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=nwg-shell#n11) instead, hence why I reported it in the [extra] package as well. I'm moving the `pacman-contrib` dependency from `nwg-shell` to `nwg-shell-config` then. :wink:

**Reply by nwg-piotr** [reply]
Excellent, thank you!

I'm working on removal of some obsolete stuff from nwg-shell-config, and necessary changes to the update notifier applet. It's going to take some more hours, so I'll make a break now, and check how what we already have installs/works.

> I also created an Arch wiki page for nwg-shell at: https://wiki.archlinux.org/title/Nwg-shell

Looks great! I'll take a closer look soon, and propose additions, if needed.

**Reply by nwg-piotr** [reply]
I installed over archinstall Minimal profile, and all went well. The config utility needs an urgent update, but I'm almost ready with this.

One thing in Wiki needs either further explanation, or a link to nwg-hello readme:

> Finally, one can [enable](https://wiki.archlinux.org/title/Enable) the greetd.service so [greetd](https://wiki.archlinux.org/title/Greetd) with the nwg-hello greeter will be started at boot.

This will come with the default agreety greeter, and may be a bit confusing.  One need to edit `/etc/greetd/config.toml` (or better its copy to `/etc/greetd/greetd.conf`), and replace:

```text
command = "agreety --cmd /bin/sh"
```

with

```text
command = "sway -c /etc/nwg-hello/sway-config"
```

or

```text
command = "Hyprland -c /etc/nwg-hello/hyprland.conf"
```

If you think this is too detailed for this Wiki page, just refer the user to https://github.com/nwg-piotr/nwg-hello#installation.

**Reply by Antiz96** [reply]
> I installed over archinstall Minimal profile, and all went well.

Good to know, same on my side it seems :)

> The config utility needs an urgent update, but I'm almost ready with this.
> 

Okay. So all we're missing now is this and the system update applet change?

> One thing in Wiki needs either further explanation, or a link to nwg-hello readme:
> [...]

Done! I added `nwg-hello` to the configuration section of every known greetd greeters to the wiki [here](https://wiki.archlinux.org/title/Greetd#nwg-hello) and linked it in the nwg-shell's wiki page like so:

"Finally, one can [enable](https://wiki.archlinux.org/title/Enable) the greetd.service (so [greetd](https://wiki.archlinux.org/title/Greetd) will be started at boot) and [set up the nwg-hello greeter](https://wiki.archlinux.org/title/Greetd#nwg-hello)."

**Reply by nwg-piotr** [reply]
Perfection!

> Okay. So all we're missing now is this and the system update applet change?

The applet is provided together with nwg-shell-config, and it seems to be ready. I'm just about to release v0.5.44.

**Reply by Antiz96** [reply]
Uh, nice! I'll package v0.5.44 right away then :)

**Reply by nwg-piotr** [reply]
Done. Hope I haven't introduced new bugs in a hurry. ;)

**Reply by Antiz96** [reply]
Well, I just pushed it to [extra-testing] so you can verify it yourself :D
Let me know :)

**Reply by nwg-piotr** [reply]
Seems to work perfectly well! :D

**Reply by Antiz96** [reply]
Cool! Does that mean we're ready to move everything to [extra]? :)

**Reply by nwg-piotr** [reply]
One issue: the `/var/cache/nwg-hello/cache.json` file should belong to the user `greeter`, but it belongs to `972`, whatever it means. I suppose this must be the reason why nwg-hello does not remember the last session data. My `greeter` user ID is 971.

**Reply by Antiz96** [reply]
I just pushed a new release of the `nwg-hello` package with a potential fix for this (you might need to reboot for it to take effect, but not sure). Can you test it?

**Reply by nwg-piotr** [reply]
Sure. Give me a minute.

**Reply by Antiz96** [reply]
Ah wait, I forgot something. Give me a sec.

**Reply by Antiz96** [reply]
Just pushed `0.2.2-4`, this one should be okay.

**Reply by nwg-piotr** [reply]
Works like a charm. We seem to be ready, or at least I hope so.

**Reply by Antiz96** [reply]
I assume the `/var/cache/nwg-hello/cache.json` file is in 755 mode now, can you confirm this?

**Reply by nwg-piotr** [reply]
It is: `-rwxr-xr-x`

**Reply by Antiz96** [reply]
Ok, I'll release 0.2.2-5 in a sec which should put it back to 644 (-rw-r--r--).

**Reply by nwg-piotr** [reply]
Sure. I'll swallow my pills and I'm back. ;)

**Reply by Antiz96** [reply]
0.2.2-5 released. If it still works as intended, we should be ready to move to [extra] I guess :)

**Reply by nwg-piotr** [reply]
> If it still works as intended

It does not. It belongs to `root` now.

Also, I forgot about `sudo` in the nwg-system-update script, so I need to release another nwg-shell-config. :/

**Reply by Antiz96** [reply]
> > If it still works as intended
> 
> It does not. It belongs to `root` now.

Indeed, I'll fix that in another release.

> 
> Also, I forgot about `sudo` in the nwg-system-update script, so I need to release another nwg-shell-config. :/

No problem, that gives me the time to fix the above :)

**Reply by nwg-piotr** [reply]
Released on my side.

**Reply by Antiz96** [reply]
Permissions of the `cache.json` file fixed with `nwg-hello 0.2.2-6` (feel free to double check) and `nwg-shell-config` v0.5.45 just released in [extra-testing].

**Reply by nwg-piotr** [reply]
The greeter cache works as expected. Couldn't test the updater script, having nothing to update, but it should work as well.

**Reply by Antiz96** [reply]
> The greeter cache works as expected. 

Nice!

> Couldn't test the updater script, having nothing to update, but it should work as well.

If you really want to test it, you can downgrade a package to trigger an update for it (e.g. `sudo pacman -U https://archive.archlinux.org/packages/t/thunar/thunar-4.18.10-2-x86_64.pkg.tar.zst`)

**Reply by nwg-piotr** [reply]
> If you really want to test it (...)

Thanks! I did so, and it worked well.

**Reply by Antiz96** [reply]
Alright! Then I'm moving everything to [extra] and removing the AUR packages :tada: 
I will also update the few links in the Arch wiki that points to the (now deprecated) AUR packages.

I'll let you know once done. Shouldn't take too long.

**Reply by nwg-piotr** [reply]
Very well. I'll update nwg-shell Wiki in the meantime, as it becomes out of date.

**Reply by Antiz96** [reply]
> Alright! Then I'm moving everything to [extra] and removing the AUR packages 🎉 I will also update the few links in the Arch wiki that points to the (now deprecated) AUR packages.
> 
> I'll let you know once done. Shouldn't take too long.

All done! `nwg-shell` is now officially in [extra] :)
Hopefully that will help for the `nwg-iso` as well ;)

**Reply by nwg-piotr** [reply]
> Hopefully that will help for the `nwg-iso` as well ;)

Definitely. It'll save me a lot of time.

Thank you for putting so much time into this. It's a pleasure working with you.

**Reply by Antiz96** [reply]
Pleasure shared ;) I'm glad we could make it!
And again, thanks for your wonderful work on the nwg-shell :)

**Comment by nwg-piotr** [comment]
@Antiz96 If I understand well, we're heading to the following situation: 


1. The nwg-shell meta-package installs all the shell components.

2. Initial config files installation (with the nwg-shell-installer script) is up to the user.

3. Login manager (nwg-hello) installation and configuration is up to the user, optionally.


Could you confirm?

**Reply by Antiz96** [reply]
Yes, basically how it was in the AUR:
Each components can be installed separately/individually, the `nwg-shell` metapkg installs them all. No scripts are run by default/automatically.

**Reply by nwg-piotr** [reply]
Very well, thank you. As for my current knowledge, I need to rewrite two scripts. Will try and do it today/tonight.

**Reply by Antiz96** [reply]
If anything, we can echo messages to users during the installation and/or have this explained in a wiki page :)

**Reply by nwg-piotr** [reply]
Or even both. Making some good wiki content will take time.

**Comment by nwg-piotr** [comment]
@Antiz96  - in case you had a spare minute. I worked for some time to get us rid of the wlr-randr dependency in all apps, and it seems we no longer need it, if we only use sway and Hyprland. Otherwise, it's listed as optional where needed (nwg-panel and Azote). Secondly, I haven't used Chromium for years, and made Firefox the default choice in nwg-iso. For consistency, let's change this in the nwg-shell package:

- remove optional dependency: wlr-randr
- remove optional dependency: chromium
- add optional dependency: firefox

- nwg-shell-install change:

> echo "WARNING: The nwg-shell-installer script expects a web browser, a file manager and a text editor to be installed before being run in order to setup keybindings properly. The suggested ones are chromium, thunar and mousepad but other ones can be installed instead."

to:

> echo "WARNING: The nwg-shell-installer script expects a web browser, a file manager and a text editor to be installed before being run in order to setup keybindings properly. The suggested ones are firefox, thunar and mousepad but other ones can be installed instead."

Thank you in advance!

**Reply by Antiz96** [reply]
All done in https://gitlab.archlinux.org/archlinux/packaging/packages/nwg-shell/-/commit/edea303cc3b8156bcbc07c853f83822ac748e6f5
I also updated the Arch wiki page accordingly: https://wiki.archlinux.org/index.php?title=Nwg-shell&diff=prev&oldid=814946

Thanks for the heads up!

**Reply by nwg-piotr** [reply]
Haha! I didn't dare ask you to change the Wiki, but you were faster. Many thanks!
